feat: Add support for "Skill Augmentation"

pacampbell opened the pull request:
feat: Add support for "Skill Augmentation"

- Add new scripting module for "Skill Augmentation"
- Locked EM4 skills behind completion of EM4
- Locked Season 2 skills and augments behind S2 BO tree
- Locked Season 3 skills and augments behind S3 BO/HO tree
- Implemented S2 tree layouts for all jobs
- Implemented new REQ, RES and handlers for Skill Augmentation
- Implemented the quest Skill Augmentation Guidance of a Renewed White Dragon
- Implemented augmentation trials for all jobs
- Updated adventure guide logic to handle when to present vocation quests better.
- Adjusted how personal quests for vocation are distributed to remove
  the need to change areas or log out to get them to show.

Thanks for help from multiple players to help us collect the skill tree data.

Co-authored-by: 𝔢𝔪𝔪𝔞 𝔢𝔳𝔢𝔯𝔤𝔞𝔯𝔡
Co-authored-by: Cold Blood
Co-authored-by: Genzyuro
Co-authored-by: あゆふぁ
Co-authored-by: 鬼頭オパーリン
Co-authored-by: GaethanT
Co-authored-by: Sapphiratelaemara
Co-authored-by: King Flipsta

# Checklist:
- [x] The project compiles
- [x] The PR targets `develop` branch

pacampbell committed:
feat: Add support for "Skill Augmentation"

- Add new scripting module for "Skill Augmentation"
- Locked EM4 skills behind completion of EM4
- Implemented tree layouts for all jobs
- Implemented new REQ, RES and handlers for Skill Augmentation
- Implemented the quest Skill Augmentation Guidance of a Renewed White Dragon
- Implemented augmentation trials for all jobs
- Updated adventure guide logic to handle when to present quests better.
- Adjusted how personal quests for vocation are distributed to remove
  the need to change areas or log out to get them to show.

Thanks for help from multiple players to help us collect the skill tree
data.

Co-authored-by: 𝔢𝔪𝔪𝔞 𝔢𝔳𝔢𝔯𝔤𝔞𝔯𝔡
Co-authored-by: Cold Blood
Co-authored-by: Genzyuro
Co-authored-by: あゆふぁ
Co-authored-by: 鬼頭オパーリン
Co-authored-by: GaethanT
Co-authored-by: Sapphiratelaemara
Co-authored-by: King Flipsta

alborrajo commented on Arrowgene.Ddon.GameServer/Characters/CharacterManager.cs:
```diff
@@ -476,5 +513,27 @@ private PacketQueue NotifyClientOfCharacterStatus(GameClient client, CharacterCo
+                .Where(x => x.SkillNo == releaseId)
+                .SelectMany(x => x.Params)
+                .Where(x => x.Lv == releaseLevel)
+                .FirstOrDefault();
```
Might be better to use Single instead so it throws an exception if there's more than one

pacampbell replied:
I added the `??  throw new ResponseError` exception pattern to these.

alborrajo commented on Arrowgene.Ddon.GameServer/Characters/CharacterManager.cs:
```diff
@@ -476,5 +513,27 @@ private PacketQueue NotifyClientOfCharacterStatus(GameClient client, CharacterCo
+                .Where(x => x.AbilityNo == releaseId)
+                .SelectMany(x => x.Params)
+                .Where(x => x.Lv == releaseLevel)
+                .FirstOrDefault();
```
ditto

RyanYappert replied:
I prefer `FirstOrDefault() ?? throw new...` because then you can at least control the exception getting thrown and add context.

alborrajo commented on Arrowgene.Ddon.GameServer/Characters/WalletManager.cs:
```diff
@@ -77,22 +77,22 @@ public CDataUpdateWalletPoint RemoveFromWallet(Character Character, WalletType T
-            {
-                return false;
-            }
+        public S2CItemUpdateCharacterItemNtc RemoveFromWalletNtc2(Character character, WalletType walletType, uint amount, DbConnection? connectionIn = null)
```
The name should probably be more descriptive, be it for this method or the other. Maybe we don't need two of them too

pacampbell replied:
I went with a Microsoft style naming when they make new functions. I think it should probably be refactored a bit a more but I didn't want to go update and break a bunch of item related handlers in this PR.

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/JobOrbTreeGetAllJobOrbElementListHandler.cs:
```diff
@@ -1,57 +1,23 @@
     public class JobOrbTreeGetJobOrbTreeGetAllJobOrbElementListHandler : GameRequestPacketHandler<C2SJobOrbTreeGetJobOrbTreeGetAllJobOrbElementListReq, S2CJobOrbTreeGetJobOrbTreeGetAllJobOrbElementListRes>
     {
-        private static readonly ServerLogger Logger = LogProvider.Logger<ServerLogger>(typeof(JobOrbTreeGetJobOrbTreeGetAllJobOrbElementListHandler));
+        private static readonly ServerLogger Logger = LogProvider.Logger<ServerLogger>(typeof(BazaarCancelHandler));
```
Why this change?

pacampbell replied:
Must be some sort of typo.

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/JobOrbTreeReleaseJobOrbElementHandler.cs:
```diff
@@ -0,0 +1,46 @@
+            });
+
+            // Notify all players of the upgrade
+            packets.AddRange(Server.CharacterManager.UpdateCharacterExtendedParamsNtc(client, client.Character));
```
Does this get sent to all players?

pacampbell replied:
Since this is a packet queue, it can contain packets for all different destinations. This particular function eventually calls `NotifyClientOfCharacterStatus` which queues up NTC packets for the player who is doing the upgrade, as well as the members in the current party.

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/SkillLearnPawnSkillHandler.cs:
```diff
@@ -17,8 +17,15 @@ public SkillLearnPawnSkillHandler(DdonGameServer server) : base(server)
+
             Pawn pawn = client.Character.PawnById(request.PawnId, PawnType.Main);
-            return Server.JobManager.UnlockSkill(Server.Database, client, pawn, request.Job, request.SkillId, request.SkillLv);
+            Server.Database.ExecuteInTransaction(connection =>
```
I assume this is done to optimize the db calls inside the job manager, right? maybe we should make the job manager methods' connection argument a non optional non nullable

pacampbell replied:
Unlock skill originally had a transaction inside. I assume to help with not opening unnecessary connections and to roll back if something failed. The general pattern so far has been to make these optional, but we can remove that if we think it's the right thing to do.

alborrajo replied:
I think so, yeah, we should have a single transaction per handler get opened

RyanYappert replied:
I'm of the opinion that managers should avoid sending packets when possible (either return a packet or a packet queue and let the handler send) and that if they need to do multiple DB operations, they should do so with an `ExecuteQuerySafe` rather than a straight transaction. That lets handlers open transactions if necessary while also making it easy if the handler forgets.

Probably violating it a lot in my own stuff, but gradual improvements.

alborrajo commented on Arrowgene.Ddon.GameServer/Characters/JobManager.cs:
```diff
@@ -159,11 +159,13 @@ public PacketQueue SetJob(GameClient client, CharacterCommon common, JobId jobId
                     otherClient.Enqueue(changeJobNotice, queue);
                 }
 
+                queue.AddRange(Server.CharacterManager.UpdateCharacterExtendedParamsNtc(client, common));
```
Gets sent to all players?

alborrajo reviewed (approved):
Amazing job

pacampbell committed:
code review feedback

- Added ?? throw to some FirstOrDefault locations when unlocking skills
  and abilities.
- Fixed a copy paste issue for the logger type in the job orb tree list
  handler.
